docs/content/sips/017-service-chaining.md

```diff
@@ -0,0 +1,122 @@
+
+Unfortunately, I don't think there's anything we can do about this, except to introduce a new `spin_lock_version`.  Doing this naively would mean existing hosts would reject the lock file _whether or not it contains prerequisites_.  We could get smart and emit a v.next lock file _only_ if the app expressed prerequisites but hoo boy.
+
+**DECISION REQUIRED:** Should we rev `spin_lock_version`?
+
+In any case, whether as a section or as an entry in `metadata`, the updated format specifies (and the Spin crate enforces) must-understand semantics for all entries under it.
+
+## Spin command line implementation
+
+## Where do invoked components run?
+
+In Spin, all components of an app run within a trigger process (typically a child Spin process, but can also be a plugin).
+
+When the calling component is on the HTTP trigger, life is (relatively) easy, because the host process has all the HTTP components - that is, all possible invocation targets - loaded and ready to go.
+
+When the calling command is on a _non-HTTP_ trigger, things are more challenging.  There are two options if we want to do this:
+
+1. We must host a large chunk of the HTTP trigger engine (and load HTTP components) in the non-HTTP trigger process. Physically this means moving the HTTP execution engine into a crate that can be shared by the `outbound-http` and `trigger-http` crates. However, this will still need a lot of the same infrastructure that the full trigger does, e.g. loading host components, permissions, etc. - I don't think it will omit much except the hooking up to the server/router. The two different ways of loading HTTP components feels like an exciting area for innovating in subtle surprises, and this is without even thinking about external trigger processes embedding different versions of the HTTP engine...
+
+2. We somehow transfer the request to the HTTP trigger for handling. This is both difficult _and_ unappealing. Let's not do it.
+
+> _Additional context:_ Today, only HTTP components can make self-requests anyway (because non-HTTP components can't infer the host URL).
+
+**For the initial release we will allow chaining only from HTTP components to HTTP components.**  This does not preclude allowing chaining from non-HTTP components in a future drop.
+
+## Streaming and concurrency
+
+It's possible for a caller to be streaming a request to a callee and the callee streaming its response at the same time.  The implementation needs to allow for this: it may just fall out for free from everything on the host side being async, but we need to test and make sure the bytes keep flowing!  Chained requests will not have hyper in the flow helpfully spawning tasks so we may end up needing to do it manually - this can be deferred to implementation.
+
```

Comment:
Failure mode: if a callee traps, what happens to the caller

Reply:
I think we should default to just returning an error to the caller. We can always allow opt-in cascading failures later if needed.